### PR TITLE
proxy v2 skip address length bytes when LOCAL command is specified

### LINKED ISSUE #4843: Proxy protocol v2 mistakenly ignores address length when LOCAL command is specified
**Jetty version**
9.4.27

**Java version**
11.0.7

**OS type/version**
Linux

**Description**
Jetty does not correctly process a proxy-protocol v2 header when the command is LOCAL and the header's address length field is non-zero.

### PR DESCRIPTION
**Pull request by lorban**
Fixes #4843

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java [hunk 1]
```diff
@@ -662,6 +662,10 @@ private void parseBodyAndUpgrade() throws IOException
                     if (LOG.isDebugEnabled())
                         LOG.debug("Proxy v2 {} {}", getEndPoint(), proxyEndPoint.toString());
                 }
+                else
+                {
+                    _buffer.position(_buffer.position() + Math.min(_buffer.remaining(), _length));
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java]: I don't think this is correct.
By this point, the buffer is guaranteed to have at least `_length` bytes in it, so `Math.min()` is not necessary.
Furthermore, I think it's wrong for LOCAL to upgrade the connection (it happens few lines below) - it was probably wrong before too.
I don't know exactly what needs to be done when receiving a LOCAL command though - perhaps just read and discard until -1. Let's discuss this.

File: jetty-server/src/test/java/org/eclipse/jetty/server/ProxyConnectionTest.java [hunk 2]
```diff
@@ -151,6 +151,39 @@ public void testIPv6V2(RequestProcessor p) throws Exception
...
+
+                // gibberish
+                "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
+            ;
+        String http = "GET /path HTTP/1.1\n" +
+            "Host: server:80\n" +
+            "Connection: close\n" +
+            "\n";
+
+        String response = p.sendRequestWaitingForResponse(TypeUtil.fromHexString(proxy), http.getBytes(StandardCharsets.US_ASCII));
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/ProxyConnectionTest.java:179]: I'm not sure you should get back a 200 OK response for a LOCAL command.

File: jetty-server/src/test/java/org/eclipse/jetty/server/ProxyProtocolTest.java [hunk 3]
```diff
@@ -193,4 +193,81 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
+
+            InputStream input = socket.getInputStream();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(input, StandardCharsets.UTF_8));
+            String response1 = reader.readLine();
+            assertTrue(response1.startsWith("HTTP/1.1 200 "));
+            while (true)
+            {
+                if (reader.readLine().isEmpty())
+                    break;
+            }
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/ProxyProtocolTest.java:253]: I'm not sure a LOCAL command should reply to HTTP requests.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/test/java/org/eclipse/jetty/server/ProxyConnectionTest.java (hunk 2 above)
Comment (lorban) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ProxyConnectionTest.java:179]: As discussed; the connection should be upgraded and handled normally even when the command is LOCAL.

File: jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java (hunk 1 above)
Comment (lorban) [reply jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java]: Regarding upgrading: as discussed; the connection should be upgraded and handled normally even when the command is LOCAL.

Regarding `Math.min()`, you're right that it's not needed as the buffer is guaranteed to contain at least `_length` bytes at that point. So I'll remove it.

**Review by sbordet — APPROVED** [review]

**Comment by joakime** [comment]
There are test failures on this branch.

**Comment by lorban** [comment]
@joakime the build looks fine to me. Could that be a browser refresh problem on your side?

**Review by joakime — APPROVED** [review]
LGTM

**Pull request merged**